### PR TITLE
HDDS-2763. Handle Datanode registration in Recon.

### PR DESCRIPTION
**Pull request by avijayanhwx**
## What changes were proposed in this pull request?
- Datanode should use Recon address to register and heartbeat to it. 
- Recon will act as a "passive" SCM to the Datanode.  Here "passive" means that the Datanode will not get any commands from this Recon SCM. 
- Integrating SCM Node Manager into Recon.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-2763

## How was this patch tested?
Manually tested using docker.
Unit tests pending.

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java [hunk 1]
```diff
@@ -64,6 +66,8 @@
 import org.apache.hadoop.net.NetUtils;
 
 import com.google.common.net.HostAndPort;
+
+import static org.apache.hadoop.hdds.recon.ReconConfigKeys.*;
```
Comment (swagle) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java]: Advice to add explicit imports.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/recon/ReconConfigKeys.java [hunk 2]
```diff
@@ -32,9 +32,11 @@ private ReconConfigKeys() {
 
   public static final String OZONE_RECON_DATANODE_ADDRESS_KEY =
       "ozone.recon.datanode.address";
+  public static final String OZONE_RECON_ADDRESS_KEY =
+      "ozone.recon.address";
```
Comment (swagle) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/recon/ReconConfigKeys.java]: Suggest rename: ozone.recon.server.address

**Comment by swagle** [comment]
We should use something like serverId on commands as blanket assurance to make sure DN does not process any commands from passive servers.

**Comment by avijayanhwx** [comment]
> We should use something like serverId on commands as blanket assurance to make sure DN does not process any commands from passive servers.

Yes, I will be doing that in a later PR, most likely in HDDS-2850.

**Comment by avijayanhwx** [comment]
Thanks for the review @swagle.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/recon/ReconConfigKeys.java (hunk 2 above)
Comment (avijayanhwx) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/recon/ReconConfigKeys.java]: I went with ozone.recon.address to keep it consistent with ozone.om.address and ozone.scm.names.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java (hunk 1 above)
Comment (avijayanhwx) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsUtils.java]: Fixed

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/recon/ReconConfigKeys.java [hunk 3]
```diff
@@ -32,9 +32,11 @@ private ReconConfigKeys() {
...
   public static final String OZONE_RECON_DATANODE_BIND_HOST_KEY =
       "ozone.recon.datanode.bind.host";
   public static final String OZONE_RECON_DATANODE_BIND_HOST_DEFAULT =
       "0.0.0.0";
-  public static final int OZONE_RECON_DATANODE_PORT_DEFAULT = 9865;
+  public static final int OZONE_RECON_DATANODE_PORT_DEFAULT = 9891;
```
Comment (anuengineer) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/recon/ReconConfigKeys.java]: This is already committed? Should we remove this change from here ?

Comment (avijayanhwx) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/recon/ReconConfigKeys.java]: Yes, it can be removed

**Pull request merged**